src/FlatSharp.Compiler.UnityStub/UnityExtensions.cs

```diff
@@ -0,0 +1,89 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using Unity.Collections;
+#if !FLATSHARP_UNITY_POLYFILLS
+using Unity.Collections.LowLevel.Unsafe;
+#endif
+
+namespace FlatSharp.Internal
+{
+#if !FLATSHARP_UNITY_POLYFILLS
+    unsafe
+#endif
+    public static class UnityExtensions
+    {
+        public static NativeArray<TElement> ReadNativeArrayBlock<TElement, TBuffer>(this TBuffer buffer, int uoffset)
+            where TElement : struct
+            where TBuffer : FlatSharp.IInputBuffer
+        {
+    #if FLATSHARP_UNITY_POLYFILLS
+            throw new NotImplementedException();
+    #else
+            // The local value stores a uoffset_t, so follow that now.
+            uoffset = uoffset + buffer.ReadUOffset(uoffset);
+
+            var mem = buffer.GetReadOnlyByteMemory(uoffset + sizeof(uint), (int)buffer.ReadUInt(uoffset));
+
+            // Create NativeArray around external memory.
+            // The assumption is this is native memory or pinned beyond 'Pin' call below. 
+            using (var pin = mem.Pin())
+            {
+                var na = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<TElement>(pin.Pointer, mem.Length, Allocator.Invalid);
+        #if ENABLE_UNITY_COLLECTIONS_CHECKS
+                NativeArrayUnsafeUtility.SetAtomicSafetyHandle(ref na, AtomicSafetyHandle.Create());
+        #endif
+                return na;
+            }
+    #endif
+        }
+    }
+
+#if !FLATSHARP_UNITY_POLYFILLS
+    unsafe
+#endif
+    public static class SpanWriterUnityExtensions
+    {
+        public static void WriteNativeArrayBlock<TElement, TSpanWriter>(
+            this TSpanWriter spanWriter,
+            Span<byte> span,
+            NativeArray<TElement> buffer,
+            int offset,
+            SerializationContext ctx)
+            where TElement : struct
+            where TSpanWriter : FlatSharp.ISpanWriter
+        {
+    #if FLATSHARP_UNITY_POLYFILLS
+            throw new NotImplementedException();
+    #else
+            int numberOfItems = buffer.Length;
+            int vectorStartOffset = ctx.AllocateVector(
+                itemAlignment: UnsafeUtility.AlignOf<TElement>(),
+                numberOfItems,
+                sizePerItem: UnsafeUtility.SizeOf<TElement>());
+
+            spanWriter.WriteUOffset(span, offset, vectorStartOffset);
+            spanWriter.WriteInt(span, numberOfItems, vectorStartOffset);
+
+            fixed (void* p = span.Slice(vectorStartOffset + sizeof(uint)))
+            {
+                UnsafeUtility.MemCpy(p, buffer.GetUnsafeReadOnlyPtr(), buffer.Length * UnsafeUtility.SizeOf<TElement>());
+            }
```

Comment:
Would some application of `MemoryMarshal.Cast` and `Span.CopyTo` work here?

Reply:
Hmm.  I didn't realize Span<T> had a (pointer, length) constructor.  There's a chance we don't have full support for Span/MemoryMarshal but if we do, yeah, seems like this could be something like `var src = new Span<TElement>(buffer.GetUnsafeReadOnlyPtr(), buffer.Length); src.CopyTo(span.Slice(...));`

Reply:
I have migrated to standard .NET BCL APIs. I'll need to check on performance of these in Unity scenario (running on Mono & IL2CPP runtimes).